libcontainer/nsenter/nsexec.c

```diff
@@ -547,13 +547,50 @@ static void update_timens_offsets(pid_t pid, char *map, size_t map_len)
 		bail("failed to update /proc/%d/timens_offsets", pid);
 }
 
+/* Check if the binary is called with the sole "init" argument by reading
+ * the /proc/self/cmdline file (as argc/argv is not available).
+ */
+static int in_init(void)
+{
+	const char expect[] = "init";
+	/* Assuming the binary path can't be longer than PATH_MAX,
+	 * read enough to check for the argument expected.
+	 */
+	char buf[PATH_MAX + sizeof(expect) + 1];
+
+	int f = open("/proc/self/cmdline", O_RDONLY | O_CLOEXEC);
+	if (f < 0)
+		bail("open /proc/self/cmdline");
+
+	int c = read(f, buf, sizeof(buf));
+	if (c < 0)
+		bail("read /proc/self/cmdline");
+
+	close(f);
+
+	char *arg = memchr(buf, '\0', c);
+	if (arg == NULL) {	/* Should never happen. */
+		return 0;
+	}
+	/* Check the size of remaining cmdline matches the expected argument. */
+	if (c - (arg - buf) != sizeof(expect) + 1) {
```

Comment:
Can't this fail if it's called like `runc init --systemd-cgroup`? (or other global options)

Reply:
`runc init` doesn't take any command-line options.

Reply:
Cool, thanks!